### PR TITLE
feat(typescript): adding typescript to protractor

### PR DESCRIPTION
**Pull request by cnishina**
Converting a 3 files over to typescript.

On the `prepublish` step, it will download the typings, transpile the files with tsc, and move the transpiled javascript files from the built directory to the lib.

Also adding scripts to package.json for `npm run tsc` and `npm run tsc:w` for transpiling.

### CODE REVIEW

**Comment by cnishina** [comment]
The linter was checking the generated javascript files from typescript. Removing the check to lib as a "fix".

File: bin/protractor [hunk 1]
```diff
@@ -2,4 +2,11 @@
 
 process.env.NODE_ENV = process.env.NODE_ENV || 'test';
 
+// Transpile code to javascript prior to running javascript
+var spawn = require('child_process').spawnSync;
+var transpile = spawn('node_modules/typescript/bin/tsc');
```
Comment (alexeagle) [inline bin/protractor]: the strategy of compiling files in the client is very different from what we do elsewhere. why not ship compiled .js files to the client?

File: lib/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+*.d.ts
```
Comment (alexeagle) [inline lib/.gitignore]: why are .d.ts files ending up here? they should not get mixed in with sources - your compiler output ought to go to a .gitignore'd folder like built/

File: tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  "compilerOptions": {
+    "target": "es5",
+    "module": "commonjs",
+    "moduleResolution": "node",
+    "sourceMap": true,
+    "declaration": true,
+    "emitDecoratorMetadata": true,
+    "experimentalDecorators": true,
+    "removeComments": false,
+    "noImplicitAny": false
```
Comment (alexeagle) [inline tsconfig.json]: add outDir: built
or similar

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{
+  "compilerOptions": {
+    "target": "es5",
+    "module": "commonjs",
+    "moduleResolution": "node",
+    "sourceMap": true,
+    "declaration": true,
+    "emitDecoratorMetadata": true,
+    "experimentalDecorators": true,
```
Comment (alexeagle) [inline tsconfig.json]: these two lines - do you use decorators or was this copy/paste from an angular2 app?

File: lib/configParser.js [hunk 5]
```diff
@@ -1,50 +1,170 @@
...
-    chromeDriver: null,
-    configDir: './',
-    plugins: [],
-    skipSourceMapSupport: false
-  };
-};
-
+catch (e) {
+}
+var ConfigParser = (function () {
```
Comment (alexeagle) [inline lib/configParser.js]: are these clang-format changes? if so it would be better in a separate commit to make this review more focused

File: lib/configParser.ts [hunk 6]
```diff
@@ -0,0 +1,230 @@
+import * as path from 'path';
+import * as glob from 'glob';
```
Comment (alexeagle) [inline lib/configParser.ts]: nit: `import {symbol}` is preferred over `import *`

File: lib/configParser.ts [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+  require('LiveScript');
+} catch (e) {
+  // Intentionally blank - ignore if LiveScript is not available.
+}
+
+export class ConfigParser {
+  public constructor() {}
+
+  // Default configuration.
+  private config_ = {
```
Comment (alexeagle) [inline lib/configParser.ts]: can there be an interface that defines the typeshape of valid configuration objects?

File: lib/configParser.ts [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+
+// LiveScript is required here to enable config files written in LiveScript.
+try {
+  require('LiveScript');
+} catch (e) {
+  // Intentionally blank - ignore if LiveScript is not available.
+}
+
+export class ConfigParser {
+  public constructor() {}
```
Comment (alexeagle) [inline lib/configParser.ts]: no need to explicitly write the no-arg constructor

File: lib/configParser.ts [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+  /**
+   * Resolve a list of file patterns into a list of individual file paths.
+   *
+   * @param {Array.<string> | string} patterns
+   * @param {=boolean} opt_omitWarnings Whether to omit did not match warnings
+   * @param {=string} opt_relativeTo Path to resolve patterns against
+   *
+   * @return {Array} The resolved file paths.
+   */
+  public static resolveFilePatterns(patterns, opt_omitWarnings, opt_relativeTo) {
```
Comment (alexeagle) [inline lib/configParser.ts]: if you add --noImplicitAny to your compiler options, you'll get typechecker errors to add types here. But you could do it later since this is an incremental migration

File: lib/configParser.ts [hunk 10]
```diff
@@ -0,0 +1,230 @@
+import * as path from 'path';
```
Comment (alexeagle) [inline lib/configParser.ts]: it's really confusing that you'll have `configParser.ts` and `configParser.js` in the same directory, seems like the output from the compiler will collide?

File: lib/configParser.ts [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+  public static resolveFilePatterns(patterns, opt_omitWarnings, opt_relativeTo) {
+    var resolvedFiles = [];
+    var cwd = opt_relativeTo || process.cwd();
+
+    patterns = (typeof patterns === 'string') ?
+        [patterns] : patterns;
+
+    if (patterns) {
+      for (var i = 0; i < patterns.length; ++i) {
+        var fileName = patterns[i];
```
Comment (alexeagle) [inline lib/configParser.ts]: nit: prefer `let` over `var`

File: lib/configParser.ts [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+   */
+  public static resolveFilePatterns(patterns, opt_omitWarnings, opt_relativeTo) {
+    var resolvedFiles = [];
+    var cwd = opt_relativeTo || process.cwd();
+
+    patterns = (typeof patterns === 'string') ?
+        [patterns] : patterns;
+
+    if (patterns) {
+      for (var i = 0; i < patterns.length; ++i) {
```
Comment (alexeagle) [inline lib/configParser.ts]: `for..of` loop?

File: lib/configParser.ts [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        union(specs, makeArray(suiteList));
+      });
+      return specs;
+    }
+
+    if (config.specs.length > 0) {
+      return config.specs;
+    }
+
+    Array.prototype.forEach.call(config.suites, function(suite) {
```
Comment (alexeagle) [inline lib/configParser.ts]: why not `config.suites.forEach()` ?

Comment (alexeagle) [inline lib/configParser.ts]: many places for fat arrows, eg
```
(suite) => { union(); }

File: lib/logger.js [hunk 14]
```diff
@@ -1,3 +1,5 @@
+/// <reference path="../typings/node/node.d.ts" />
```
Comment (alexeagle) [inline lib/logger.js]: prefer to give node.d.ts to the compiler as an input, don't reference in your code

Comment (alexeagle) [inline lib/logger.js]: also this is a .js file, so the comment does nothing? unless you're using TS 1.8 support
https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#including-js-files-with---allowjs

Comment (alexeagle) [inline lib/logger.js]: oh! you are checking in generated code!! don't do that :)

File: tsd.json [hunk 15]
```diff
@@ -0,0 +1,21 @@
+{
+  "version": "v4",
+  "repo": "borisyankov/DefinitelyTyped",
```
Comment (alexeagle) [inline tsd.json]: looks old, should be DefinitelyTyped/DefinitelyTyped

File: typings/glob/glob.d.ts [hunk 16]
```diff
@@ -0,0 +1,112 @@
+// Type definitions for Glob 5.0.10
```
Comment (alexeagle) [inline typings/glob/glob.d.ts]: checking in the typings can be a valid thing to do, because the github connection can be flaky...
but in our other projects, we run `typings install ...` in an npm postinstall script and don't check in the files

File: typings/tsd.d.ts [hunk 17]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (alexeagle) [inline typings/tsd.d.ts]: opt: .gitignore the typings folder

Comment (cnishina) [inline typings/tsd.d.ts]: Done.

File: tsd.json (hunk 15 above)
Comment (cnishina) [inline tsd.json]: Removing tsd.json for typings.json.

File: typings/glob/glob.d.ts (hunk 16 above)
Comment (cnishina) [inline typings/glob/glob.d.ts]: Done. Added  `typings install` to post install script in package.json.

File: lib/configParser.ts (hunk 13 above)
Comment (cnishina) [inline lib/configParser.ts]: * Replaced `Array.prototype.forEach` with `config.suites.forEach`.
* Went over typescript, cleaned up all with fat arrows.

File: lib/configParser.ts (hunk 12 above)
Comment (cnishina) [inline lib/configParser.ts]: Done.

File: lib/configParser.ts (hunk 11 above)
Comment (cnishina) [inline lib/configParser.ts]: Cleaned up all files for `let` over use of `var`

File: lib/configParser.ts (hunk 9 above)
Comment (cnishina) [inline lib/configParser.ts]: Fixed all implicit any and added types.

File: lib/configParser.ts (hunk 8 above)
Comment (cnishina) [inline lib/configParser.ts]: Removed.

File: lib/configParser.ts (hunk 7 above)
Comment (cnishina) [inline lib/configParser.ts]: Done, created a Config interface.

File: tsconfig.json (hunk 4 above)
Comment (cnishina) [inline tsconfig.json]: Removed.

File: lib/.gitignore (hunk 2 above)
Comment (cnishina) [inline lib/.gitignore]: Moved *.d.ts to built/ directory and added built/ to the .gitignore

File: tsconfig.json (hunk 3 above)
Comment (cnishina) [inline tsconfig.json]: Added outDir: built

File: lib/.gitignore [hunk 18]
```diff
@@ -0,0 +1,4 @@
+# TODO: remove this file when all files have been replaced with typescript
+configParser.js
```
Comment (alexeagle) [inline lib/.gitignore]: do you still need this? seems like the file is now removed by this change

File: lib/configParser.ts [hunk 19]
```diff
@@ -0,0 +1,258 @@
+import { resolve, dirname } from 'path';
```
Comment (alexeagle) [inline lib/configParser.ts]: consider running all your .ts files through clang-format so I don't need to nitpick about whitespace :)

You can see in other repos (eg. angular/ts2dart) how we've hooked up clang-format so it's convenient to run and enforced by the build.

File: lib/configParser.ts [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+  // Intentionally blank - ignore if LiveScript is not available.
+}
+
+export interface Config {
+  specs: Array<string>;
+  multiCapabilities: Array<any>;
+  rootElement: string;
+  allScriptsTimeout: number;
+  getPageTimeout: number;
+  params: any;
```
Comment (alexeagle) [inline lib/configParser.ts]: I am surprised that none of these are marked optional. Doesn't it fail to typecheck if you create a Config object that is missing one of the properties?

File: lib/configParser.ts [hunk 21]
```diff
@@ -0,0 +1,258 @@
...
+      seleniumSessionId: null,
+      mochaOpts: {
+        ui: 'bdd',
+        reporter: 'list'
+      },
+      chromeDriver: null,
+      configDir: './',
+      plugins: [],
+      skipSourceMapSupport: false,
+      suites: null,
```
Comment (alexeagle) [inline lib/configParser.ts]: seems like this property wants to be optional, for example

File: lib/configParser.ts [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+      suite: null
+    };
+
+  }
+
+  /**
+   * Resolve a list of file patterns into a list of individual file paths.
+   *
+   * @param {Array.<string> | string} patterns
```
Comment (alexeagle) [inline lib/configParser.ts]: what's your long-term plan with the JSDoc type annotations? will these be kept in sync with the typescript typings? Is anything type-checking them? Does the docgen use them?

Seems to me these should just be removed in this change so later maintainers don't have to deal with them.

File: lib/configParser.ts [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+   * @param {=boolean} opt_omitWarnings Whether to omit did not match warnings
+   * @param {=string} opt_relativeTo Path to resolve patterns against
+   *
+   * @return {Array} The resolved file paths.
+   */
+  public static resolveFilePatterns(patterns: Array<string>|string,
+      opt_omitWarnings: boolean, opt_relativeTo: string) {
```
Comment (alexeagle) [inline lib/configParser.ts]: missing the optionality (should have a `?`)

File: lib/configParser.ts [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
+    let resolvedFiles: Array<string> = [];
+    let cwd = opt_relativeTo || process.cwd();
+
+    patterns = (typeof patterns === 'string') ?
+        [patterns] : patterns;
+
+    if (patterns) {
+      for (let fileName of patterns) {
+        let matches = sync(fileName, {cwd: cwd});
```
Comment (alexeagle) [inline lib/configParser.ts]: I believe ES6 allows this to be just `{cwd}` as a shorthand for property and value being the same identifier
http://people.mozilla.org/~jorendorff/es6-draft.html#sec-object-initializer

try that out?

File: lib/configParser.ts [hunk 25]
```diff
@@ -0,0 +1,258 @@
...
+  /**
+   * Public function specialized towards merging in a file's config
+   *
+   * @public
+   * @param {String} filename
+   */
+  public addFileConfig(filename: string) {
+    try {
+      if (!filename) {
+        return this;
```
Comment (alexeagle) [inline lib/configParser.ts]: should this method have a declared return type?

File: lib/driverProviders/sauce.js [hunk 26]
```diff
@@ -5,7 +5,7 @@
  */
 
 var util = require('util'),
-    log = require('../logger.js'),
+    log = require('../logger.js').Logger,
```
Comment (alexeagle) [inline lib/driverProviders/sauce.js]: If you like, you can avoid this change by continuing to use default export for the Logger symbol.
https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#11.3.4.2

File: lib/logger.ts [hunk 27]
```diff
@@ -0,0 +1,38 @@
+/**
+ * Utility functions for command line output logging from Protractor.
+ * May be used in different processes, since the launcher spawns
+ * child processes for each runner.
+ *
+ * All logging directly from Protractor, its driver providers, or runners,
+ * should go through this file so that it can be customized.
+ */
+export module Logger {
```
Comment (alexeagle) [inline lib/logger.ts]: this looks like old TypeScript syntax for a "internal module" which was replaced by namespace in https://github.com/Microsoft/TypeScript/issues/2159

https://basarat.gitbooks.io/typescript/content/docs/project/namespaces.html

File: package.json [hunk 28]
```diff
@@ -45,7 +47,10 @@
   },
   "main": "lib/protractor.js",
   "scripts": {
-    "pretest": "./scripts/pretest.sh",
+    "prepublish": "./node_modules/.bin/typings install && ./node_modules/typescript/bin/tsc && cp built/*.js lib/",
```
Comment (alexeagle) [inline package.json]: I think I was wrong about this. You have typings in devDependencies, so the typings install command belongs in postinstall script (otherwise your teammates won't be able to compile until they attempt to publish!)

File: tsconfig.json [hunk 29]
```diff
@@ -0,0 +1,24 @@
...
+    "removeComments": false,
+    "noImplicitAny": true,
+    "outDir": "built/"
+  },
+  "exclude": [
+    "typings",
+    "node_modules",
+    "testapp"
+  ],
+  "files": [
```
Comment (alexeagle) [inline tsconfig.json]: don't include both exclude and files.

https://github.com/Microsoft/TypeScript/wiki/tsconfig.json#details says
The "files" property cannot be used in conjunction with the "exclude" property. If both are specified then the "files" property takes precedence.

files will probably get you into trouble when you have to list every single test file, so just use exclude and let the compiler find all `.ts` files under this directory.

File: lib/configParser.ts (hunk 22 above)
Comment (juliemr) [inline lib/configParser.ts]: We should indeed remove them.

File: package.json (hunk 28 above)
Comment (juliemr) [inline package.json]: Would this override what's currently in lib? I think it would be cleaner to just move everything we need into `built` and publish from there.

File: lib/.gitignore (hunk 18 above)
Comment (cnishina) [inline lib/.gitignore]: This is still needed. On the prepublish step, the transpiled javascript files are being copied into the lib folder.

Comment (alexeagle) [inline lib/.gitignore]: I see, because the published app references lib/file.js in package.json
maybe it should reference built/file.js instead, like Julie suggests

File: lib/driverProviders/sauce.js (hunk 26 above)
Comment (cnishina) [inline lib/driverProviders/sauce.js]: In logger.js, looks like it would require changing the module to a class to use `export default`. Trying that out, it looks like `default` would make calls like `log.set(` change to `log.default.set(` for javascript files like this one.

File: lib/logger.ts (hunk 27 above)
Comment (cnishina) [inline lib/logger.ts]: When changing to `export namespace Logger {`, it appears that when doing a `var log = require('./logger')` would change the calls like `log.set` to  `log.Logger.set(` for javascript files that use it. Please advise.

File: tsconfig.json (hunk 29 above)
Comment (cnishina) [inline tsconfig.json]: Fixed.

File: lib/configParser.ts (hunk 20 above)
Comment (cnishina) [inline lib/configParser.ts]: Done. Added variables as optional to the Config object.

File: lib/configParser.ts (hunk 21 above)
Comment (cnishina) [inline lib/configParser.ts]: Done. Removed null from config_ since they are now optional.

File: lib/driverProviders/sauce.js (hunk 26 above)
Comment (alexeagle) [inline lib/driverProviders/sauce.js]: No, there is another syntax to add an export default at the end of the file, choosing which symbol from the scope should be the default export.
So TypeScript should be able to express the logger.ts file so that the usage sites are unchanged.

File: lib/logger.ts (hunk 27 above)
Comment (alexeagle) [inline lib/logger.ts]: look around for another way to export - if you are stuck please ping me on IM and I'll patch your change locally and confirm if I'm right :)

File: lib/configParser.ts (hunk 19 above)
Comment (cnishina) [inline lib/configParser.ts]: Done. Added clang-format to part of the prepublish / pretest step.

File: lib/configParser.ts (hunk 22 above)
Comment (cnishina) [inline lib/configParser.ts]: Per a discussion we had today, we will remove them in the future. Issue track for this is: #2930.

**Comment by cnishina** [comment]
Added jslint check back for `lib/` since transpiled files are now in `built/`

File: .gitignore [hunk 30]
```diff
@@ -12,3 +12,6 @@ xmloutput*
 npm-debug.log
 
 *.swp
+
+typings/
+built/
```
Comment (juliemr) [inline .gitignore]: built/ is already at the top from the other PR, remove this line

File: gulpfile.js [hunk 31]
```diff
@@ -1,39 +1,59 @@
...
 var gulp = require('gulp');
+var clangFormat = require('clang-format');
+var gulpFormat = require('gulp-clang-format');
 var runSequence = require('run-sequence');
-var spawnSync = require('child_process').spawnSync;
+var spawn = require('child_process').spawn;
+
+var runSpawn = function(done, task, opt_arg) {
+  var child = spawn(task, opt_arg);
+  child.stdout.on('data', function(data) {
```
Comment (juliemr) [inline gulpfile.js]: Instead of manually using console.log, can we use child_process stdio options? (https://nodejs.org/dist/latest-v4.x/docs/api/child_process.html#child_process_options_stdio) I think 'inherit' does what you want here.

File: testapp/ng2/app/async/async.component.js [hunk 32]
```diff
@@ -29,7 +29,7 @@ System.register(['angular2/core', 'angular2/common', 'angular2/src/facade/async'
                     this.val3 = 0;
```
Comment (juliemr) [inline testapp/ng2/app/async/async.component.js]: Changes here seem unrelated, can this be a separate PR?

File: tsconfig.json [hunk 33]
```diff
@@ -0,0 +1,20 @@
...
+    "declaration": true,
+    "removeComments": false,
+    "noImplicitAny": false,
+    "outDir": "built/"
+  },
+  "exclude": [
+    "built",
+    "node_modules",
+    "testapp",
+    "typings/browser",
```
Comment (juliemr) [inline tsconfig.json]: Is there a reason not to just exclude all of 'typings'?

**Comment by juliemr** [comment]
Just a couple cleanup comments, otherwise looks good!

File: tsconfig.json (hunk 33 above)
Comment (cnishina) [inline tsconfig.json]: There needs to be a reference to the main.d.ts file.

File: testapp/ng2/app/async/async.component.js (hunk 32 above)
Comment (cnishina) [inline testapp/ng2/app/async/async.component.js]: This should not be part of the PR. Will remove this.

File: gulpfile.js (hunk 31 above)
Comment (cnishina) [inline gulpfile.js]: Done.

File: .gitignore (hunk 30 above)
Comment (cnishina) [inline .gitignore]: Done.

**Comment by cnishina** [comment]
Reference feature request: #1203
